### PR TITLE
Fog: Add fog_encryption option; honor url protocol when generating expiring_url

### PR DESCRIPTION
**Pull request by stereoscott**
Two features to enhance security when using fog:
1. Add a fog_encryption option pass encryption option to fog. This follows a similar pattern to the s3 provider implementation, and has the added benefit of relying on fog to set the proper header for whatever provider may be in use (see [fog's implementation here](https://github.com/fog/fog/blob/41fddec67d6f4ee02e3b0f9d6de48b12e8ecaa37/lib/fog/aws/models/storage/file.rb#L188)).
2. Recognize the url protocol when generating an expiring url instead of always using `get_http_url` which  could result in duplicate protocols when using fog_host and prevent secure urls. This addresses https://github.com/thoughtbot/paperclip/issues/1112.

### CODE REVIEW

**Comment by maclover7** [comment]
Hi @stereoscott ! Is this still an issue? If so, can you please rebase your PR, fix Hound notes, and make sure everything is passing? Thanks!

File: spec/paperclip/storage/fog_spec.rb [hunk 1]
```diff
@@ -337,12 +343,16 @@ def custom_method
...
+          expect(@dummy.avatar.url).to match(
+            /^https:\/\/papercliptests.s3.amazonaws.com\/avatars\/5k.png/
+          )
         end
 
-        it "provides an url that expires in subdomain style" do
-          assert_match(/^http:\/\/papercliptests.s3.amazonaws.com\/avatars\/5k.png.+Expires=.+$/, @dummy.avatar.expiring_url)
+        it "provides a url in subdomain style that expires" do
+          expect(@dummy.avatar.expiring_url).to match(
+            /^https:\/\/papercliptests.s3.amazonaws.com\/avatars\/5k.png.+Expires=.+$/
```
Comment (stereoscott) [inline spec/paperclip/storage/fog_spec.rb]: How might one shorten this line?

File: spec/paperclip/storage/fog_spec.rb [hunk 2]
```diff
@@ -386,11 +396,15 @@ def custom_method
         end
 
         it "provides an url in folder style" do
-          assert_match(/^https:\/\/s3.amazonaws.com\/this_is_invalid\/avatars\/5k.png\?\d*$/, @dummy.avatar.url)
+          expect(@dummy.avatar.url).to match(
+            /^https:\/\/s3.amazonaws.com\/this_is_invalid\/avatars\/5k.png\?\d*$/
```
Comment (stereoscott) [inline spec/paperclip/storage/fog_spec.rb]: Do we really want to break up this regex? How?

**Comment by seanjoo** [comment]
@stereoscott @jyurek any chance this will be merged anytime soon?

**Comment by sidraval** [comment]
We are in the process of deprecating paperclip, and so are closing PRs that aren't immediate bug fixes or documentation updates. Thanks for your contribution, we regret not tending to it!

**Pull request closed without merging**